pyro/distributions/transforms/spline.py

```diff
@@ -0,0 +1,352 @@
+# This implementation is adapted in part from:
+# * https://github.com/tonyduan/normalizing-flows/blob/master/nf/flows.py; and,
+# * https://github.com/hmdolatabadi/LRS_NF/blob/master/nde/transforms/nonlinearities.py,
+# under the MIT license.
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from torch.distributions import constraints
+
+from pyro.distributions.torch_transform import TransformModule
+from pyro.distributions.util import copy_docs_from
+
+# TODO: Move the spline functions to a pyro.numerics or torch.numerics module?
+# This would be a good place for e.g. numerical ODE solvers as well...
+
+
+def search_sorted(bin_locations, inputs, eps=1e-6):
+    """
+    Searches for which bin an input belongs to (in a way that is parallelizable and amenable to autodiff)
+    """
+    bin_locations[..., -1] += eps
+    return torch.sum(
+        inputs[..., None] >= bin_locations,
+        dim=-1
+    ) - 1
+
+
+def select_bins(x, idx):
+    """
+    Performs gather to select the bin in the correct way on batched inputs
+    """
+    idx = idx.clamp(min=0, max=x.size(-1) - 1)
+    extra_dims = len(idx.shape) - len(x.shape)
+    expanded_x = x.expand(idx.shape[:extra_dims] + (-1,) * len(x.shape))
+    return expanded_x.gather(-1, idx).squeeze(-1)
+
+
+def calculate_knots(lengths, lower, upper):
+    """
+    Given a tensor of unscaled bin lengths that sum to 1, plus the lower and upper limits,
+    returns the shifted and scaled lengths plus knot positions
+    """
+
+    # Cumulative widths gives x (y for inverse) position of knots
+    knots = torch.cumsum(lengths, dim=-1)
+
+    # Pad left of last dimension with 1 zero to compensate for dim lost to cumsum
+    knots = F.pad(knots, pad=(1, 0), mode='constant', value=0.0)
+
+    # Translate [0,1] knot points to [-B, B]
+    knots = (upper - lower) * knots + lower
+
+    # Convert the knot points back to lengths
+    # NOTE: Are following two lines a necessary fix for accumulation (round-off) error?
+    knots[..., 0] = lower
+    knots[..., -1] = upper
+    lengths = knots[..., 1:] - knots[..., :-1]
+
+    return lengths, knots
+
+
+def monotonic_rational_spline(inputs, widths, heights, derivatives, lambdas,
+                              inverse=False,
+                              bound=3.,
+                              min_bin_width=1e-3,
+                              min_bin_height=1e-3,
+                              min_derivative=1e-3,
+                              min_lambda=0.025):
+    """
+    Calculating a monotonic rational spline (linear for now) or its inverse, plus the log(abs(detJ)) required
+    for normalizing flows.
+    NOTE: I omit the docstring with parameter descriptions for this method since it is not considered "public" yet!
```

Comment:
please prefix private methods with underscore, or demote docstring to a comment.

Now that Pyro is 1.0 we need to take special care to ensure all public interface changes are backwards compatible. The easy way to do this is simply not publish methods. Following Python convention we can indicate privacy with leading underscores, but Pyro also restricts stability to "documented methods" (see [RELEASE_MANAGEMENT.md](https://github.com/pyro-ppl/pyro/blob/dev/RELEASE-MANAGEMENT.md)). You can also mark this method EXPERIMENTAL in the docstring. ditto for other method in this file that you plan to either change or move.

Reply:
I'll leave as private with an underscore for now